sdk/dotnet/Pulumi/Serialization/Converter.cs

```diff
@@ -0,0 +1,264 @@
+                    return Activator.CreateInstance(targetType);
+
+                // for all other types, can just return the null value right back out as a legal
+                // reference type value.
+                return null;
+            }
+
+            // We're not null and we're converting to Nullable<T>, just convert our value to be a T.
+            if (targetIsNullable)
+                return ConvertObject(context, val, targetType.GenericTypeArguments.Single());
+
+            if (targetType == typeof(string))
+                return EnsureType<string>(context, val);
+
+            if (targetType == typeof(bool))
+                return EnsureType<bool>(context, val);
+
+            if (targetType == typeof(double))
+                return EnsureType<double>(context, val);
+
+            if (targetType == typeof(int))
+                return (int)EnsureType<double>(context, val);
+
+            if (targetType == typeof(Asset))
+                return EnsureType<Asset>(context, val);
+
+            if (targetType == typeof(Archive))
+                return EnsureType<Archive>(context, val);
+
+            if (targetType == typeof(AssetOrArchive))
+                return EnsureType<AssetOrArchive>(context, val);
+
+            if (targetType.IsConstructedGenericType)
+            {
+                if (targetType.GetGenericTypeDefinition() == typeof(ImmutableArray<>))
+                    return ConvertArray(context, val, targetType);
+                
+                if (targetType.GetGenericTypeDefinition() == typeof(ImmutableDictionary<,>))
+                    return ConvertDictionary(context, val, targetType);
+                
+                throw new InvalidOperationException(
+                    $"Unexpected generic target type {targetType.FullName} when deserializing {context}");
+            }
+
+            if (targetType.GetCustomAttribute<OutputTypeAttribute>() == null)
+                throw new InvalidOperationException(
+                    $"Unexpected target type {targetType.FullName} when deserializing {context}");
+
+            var constructor = GetPropertyConstructor(targetType);
+            if (constructor == null)
+                throw new InvalidOperationException(
+                    $"Expected target type {targetType.FullName} to have [PropertyConstructor] constructor when deserializing {context}");
+
+            var dictionary = EnsureType<ImmutableDictionary<string, object>>(context, val);
+
+            var constructorParameters = constructor.GetParameters();
+            var arguments = new object?[constructorParameters.Length];
+
+            for (int i = 0, n = constructorParameters.Length; i < n; i++)
+            {
+                var parameter = constructorParameters[i];
+
+                // Note: TryGetValue may not find a value here.  That can happen for things like
+                // unknown vals.  That's ok.  We'll pass that through to 'Convert' and will get the
```

Comment:
Have you given any thought to how first-class-unknowns will work in .NET?

Reply:
No.  Adding to list.

Reply:
From slack discussion. We likely need to just represent everything as nullable (either ref or val) in the final shape we produce.  For most things this is not too bad (though it's unfortunate that primitive types would become Nullable<T>).  

We then doc that by default if anything came through as null in preview that we won't want run the applies so you normally won't have to deal with nulls.  And if you do want to potentially handle that, you pass in a parameter saying you want to run, and you write your code defensively.  This isn't too bad in c# since there are constructs like `??` and `?.` and so on.